library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java

```
@@ -384,20 +394,20 @@ public void put(Slice key, long timeBucket, Slice value)
     {
       BucketedValue bucketedValue = flash.get(key);
       if (bucketedValue == null) {
-        bucketedValue = new BucketedValue();
+        bucketedValue = new BucketedValue(timeBucket, value);
         flash.put(key, bucketedValue);
-        sizeInBytes.getAndAdd(key.length);
-        sizeInBytes.getAndAdd(Long.SIZE);
-      }
-      if (timeBucket >= bucketedValue.getTimeBucket()) {
-
-        int inc = null == bucketedValue.getValue() ? value.length : value.length - bucketedValue.getValue().length;
-        sizeInBytes.getAndAdd(inc);
-        bucketedValue.setTimeBucket(timeBucket);
-        bucketedValue.setValue(value);
+        sizeInBytes.getAndAdd(key.length + value.length + Longs.BYTES);
       } else {
-        LOG.warn("ignoring {} {} {}; existing {} {}", key, value, timeBucket,
-            bucketedValue.getValue(), bucketedValue.getTimeBucket());
+        if (timeBucket >= bucketedValue.getTimeBucket()) {
+
+          int inc = null == bucketedValue.getValue() ? value.length : value.length - bucketedValue.getValue().length;
+          sizeInBytes.getAndAdd(inc);
+          bucketedValue.setTimeBucket(timeBucket);
+          bucketedValue.setValue(value);
+        } else {
+          LOG.warn("ignoring {} {} {}; existing {} {}", key, value, timeBucket,
```

Shouldn't this raise an exception instead of logging if the put is not applied?

---

Seemed like out of order updates to an entry can be frequent. Ignoring them seemed better.
However, I am not sure after recent discussions. Maybe we should throw an exception, so the clients will have to handle it.
Copying @davidyan74 as well.

---

Can the caller check the condition before calling put()? 
IMO, a call to put() should always do what it's expected to do (creating an entry or overwriting an entry).

---

If this method was named "putIfMostRecent" or something like that, I would not have an issue of it ignoring a call quietly based on a certain condition.

---

Caller can do a get before put, however get returns only a Slice.
We can change it to return `BucketedValue` however that is part of the interface and the change may touch several classes using ManagedState

---

Actually, caller doesn't know timeBucket for the new entry as well. TimeBucket is created inside managed state so the caller can't even compare it.

---

Updated the pull request to throw an exception.